File: src/main/java/org/xyp/project/todoapp/core/txop/impl/RealTx.java
package org.xyp.project.todoapp.core.txop.impl;

import java.util.function.Supplier;

public interface RealTx {


    <T> T doReturnInTx(Supplier<T> supplier);

    <T> T doReturnInNewTx(Supplier<T> supplier);

    void doRunInTx(Runnable runner);

    void doRunInNewTx(Runnable runner);
}

File: src/main/kotlin/org/xyp/todoapp/core/txop/impl/RealTx.kt
package org.xyp.todoapp.core.txop.impl

import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class RealTx {

    @Transactional(value = Transactional.TxType.REQUIRED, rollbackOn = [Exception::class])
    fun <T> doReturnInTx(supplier: () -> T?): T? {
        return supplier()
    }

    @Transactional(value = Transactional.TxType.REQUIRES_NEW, rollbackOn = [Exception::class])
    fun <T> doReturnInNewTx(supplier: () -> T?): T? {
        return supplier()
    }

    @Transactional(value = Transactional.TxType.REQUIRED, rollbackOn = [Exception::class])
    fun doRunInTx(runner: () -> Unit) {
        runner()
    }

    @Transactional(value = Transactional.TxType.REQUIRES_NEW, rollbackOn = [Exception::class])
    fun doRunInNewTx(runner: () -> Unit) {
        runner()
    }
}